report/sections/3-typescript.tex

```diff
@@ -0,0 +1,190 @@
+  constructor(head: A) {
+    this.head = head;
+    this.tail = new List();
+  }
+}
+\end{lstlisting}
+
+In any nominally typed language, it would not be possible to assign a \texttt{NonEmptyList} to any variable that is not also a \texttt{NonEmptyList}.
+In TypeScript this restriction does not apply. 
+Any instance of a class can be assigned a variable with any type as long as the two are structurally in a subtype relationship.
+This means that the following snippet is perfectly legal:
+
+\begin{lstlisting}
+class SignletonList<A> {
+  head: A;
+}
+
+const s: SingletonList<number> = 
+  new NonEmptyList(42);
+\end{lstlisting}
+
+Since names are irrelevant to the type checker, the following is also valid because the type of 
+\texttt{s} is structurally equivalent to \texttt{NonEmptyList<number>}.
+\bigskip
+\begin{lstlisting}
+const s: { 
+  head: number, tail: List<number> 
+} = new NonEmptyList(42);
+\end{lstlisting}
+
+This situation remains unchanged when we add 'methods' to the \texttt{NonEmptyList} class:
+
+\begin{lstlisting}
+class NonEmptyList<A> {
+  head: A;
+  tail: List<A>;
+  ...
+  size(): number {
+    return 1 + tail.size();
+  }
+}
+
+const s: { 
+  head: number, tail: List<number>,
+  size: () => number  
+} = new NonEmptyList(42);
+\end{lstlisting}
+
+From the above snippet we can see that classes are simply sugar for named object types.
+The fields are directly translated to object members, and class methods are translated
+to object members that have a function type corresponding to the method signature.
+A consequence of this is that class methods are first class values.
+Class constructors are translated to a special type of function outside the class scope that has the exact name of the class:
+
+\begin{lstlisting}
+class NonEmptyList<A> {
+  ...
+  constructor(head: A) {
+    ...
+  }
+}
+const a: typeof NonEmptyList = NonEmptyList
+const xs: NonEmptyList<number> = new a(42)
+\end{lstlisting}
+
+From this it appears that classes live in both the value namespace as well as the type namespace,
+because an instance of \texttt{NonEmptyList} ($xs$) can be created from a runtime value ($a$) on the last line.
+In reality, $a$ represents the constructor for \texttt{NonEmptyList} rather than the class itself.
+We can therefore conclude that class constructors are first class values just like class methods.
+The type of a constructor must different from the type of a function because the former can only be called with the \texttt{new} keyword.
+The \texttt{typeof} keyword in the definition of $a$ is used in this case to signal that $a$ is a constructor.
+Since constructors do not live inside the class scope, they have no impact type checking when handling instances of a class.
+\bigskip
+Inheritance is also handled in a straightforward manner.
+When checking whether an instance of a class can be assigned to an object type, the compiler
+simply takes all class members and all members of any parent class into account.
```

Comment:
Add here that the fields of the value should be a superset of the fields specified in the object type